# Google Play Store Data Cleaning

In [46]:
PlayStoreData = "./raw/Play Store Data.csv"
UserReviews = "./raw/User Reviews.csv"

## **Phase 1 :** Setup & Loading

In [47]:
import pandas as pd
import numpy as np

In [48]:
playstore = pd.read_csv(PlayStoreData)
userreviews = pd.read_csv(UserReviews)

## **Phase 2 :** Data Exploration

### Getting info about datasets

In [49]:
playstore.shape

(10841, 13)

In [50]:
userreviews.shape

(64295, 5)

In [51]:
playstore.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


In [52]:
userreviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64295 entries, 0 to 64294
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     64295 non-null  object 
 1   Translated_Review       37427 non-null  object 
 2   Sentiment               37432 non-null  object 
 3   Sentiment_Polarity      37432 non-null  float64
 4   Sentiment_Subjectivity  37432 non-null  float64
dtypes: float64(2), object(3)
memory usage: 2.5+ MB


In [53]:
playstore.head()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up


In [54]:
userreviews.head()

,App,Translated_Review,Sentiment,Sentiment_Polarity,Sentiment_Subjectivity
0,10 Best Foods for You,I like eat delicious food. That's I'm cooking ...,Positive,1.00,0.533333
1,10 Best Foods for You,This help eating healthy exercise regular basis,Positive,0.25,0.288462
2,10 Best Foods for You,NaN,NaN,NaN,NaN
3,10 Best Foods for You,Works great especially going grocery store,Positive,0.40,0.875000
4,10 Best Foods for You,Best idea us,Positive,1.00,0.300000


### **Getting Missing Values**

In [55]:
userreviews.isnull().sum()

App                           0
Translated_Review         26868
Sentiment                 26863
Sentiment_Polarity        26863
Sentiment_Subjectivity    26863
dtype: int64

In [56]:
playstore.isnull().sum()

App                  0
Category             0
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       1
Genres               0
Last Updated         0
Current Ver          8
Android Ver          3
dtype: int64

### **Getting Duplicate Values**

In [57]:
userreviews.duplicated().sum()

np.int64(33616)

In [58]:
playstore.duplicated().sum()

np.int64(483)

### **Describing Tables**

In [59]:
userreviews.describe()

,Sentiment_Polarity,Sentiment_Subjectivity
count,37432.000000,37432.000000
mean,0.182146,0.492704
std,0.351301,0.259949
min,-1.000000,0.000000
25%,0.000000,0.357143
50%,0.150000,0.514286
75%,0.400000,0.650000
max,1.000000,1.000000


In [60]:
playstore.describe()

,Rating
count,9367.000000
mean,4.193338
std,0.537431
min,1.000000
25%,4.000000
50%,4.300000
75%,4.500000
max,19.000000


### **Columns Detail**

In [61]:
playstore.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')

In [62]:
userreviews.columns

Index(['App', 'Translated_Review', 'Sentiment', 'Sentiment_Polarity',
       'Sentiment_Subjectivity'],
      dtype='object')

### **Conclusion**

Based on these initial checks, I cocluded that both cleaning and transformation is required to the data before they can effectly analyzed
- **Data Type Conversion :** Several columns in `playstore` need to be converted to approperiate numeric or datetime types. This process we require the removal of non-numberic values and handling outliers.
- **Handling Missing Values:**
  - In `playstore` we have *Rating* column has *1474* missing values and columns like `Type`, `Content Rating`, `Current Ver`, and `Android Ver` has very small amount of missings
  - In the `userreviews` dataset, `Translated_Review`, `Sentiment`, `Sentiment_Polarity`, and` Sentiment_Subjectivity` each have over 26,800 missing values which will directly effect the sentiment analysis.
- **Duplicates:**
  - I found `483` duplicate rows in the `playstore` dataset.
  - A significant `33616` duplicate rows were identified in the `userreviews` dataset.

## **Phase 3 :** Data Cleaning & Transformation

### Working with `Installs` feature in **playstore** dataset

In [63]:
playstore['Installs'].value_counts()

Installs
1,000,000+        1579
10,000,000+       1252
100,000+          1169
10,000+           1054
1,000+             907
5,000,000+         752
100+               719
500,000+           539
50,000+            479
5,000+             477
100,000,000+       409
10+                386
500+               330
50,000,000+        289
50+                205
5+                  82
500,000,000+        72
1+                  67
1,000,000,000+      58
0+                  14
0                    1
Free                 1
Name: count, dtype: int64

In [64]:
playstore['Installs'] = playstore['Installs'].str.replace("Free", "0").str.replace('[+,]', "", regex=True).astype(np.int64)

In [65]:
playstore['Installs'].value_counts()

Installs
1000000       1579
10000000      1252
100000        1169
10000         1054
1000           907
5000000        752
100            719
500000         539
50000          479
5000           477
100000000      409
10             386
500            330
50000000       289
50             205
5               82
500000000       72
1               67
1000000000      58
0               16
Name: count, dtype: int64

### Working with `Price` feature in **playstore** dataset

In [66]:
playstore["Price"].value_counts()

Price
0           10040
$0.99         148
$2.99         129
$1.99          73
$4.99          72
            ...  
$394.99         1
$1.26           1
Everyone        1
$1.20           1
$1.04           1
Name: count, Length: 93, dtype: int64

In [67]:
playstore["Price"] = playstore["Price"].str.replace("Everyone", "0").str.replace("$", "").astype(np.float16)

In [68]:
playstore["Price"].value_counts()

Price
0.000000      10041
0.990234        148
2.990234        129
1.990234         73
4.988281         72
              ...  
3.609375          1
395.000000        1
1.259766          1
1.200195          1
1.040039          1
Name: count, Length: 91, dtype: int64

### Working with `Size` feature in **playstore** dataset

In [69]:
playstore["Size"].value_counts().sort_values()

Size
811k                     1
688k                     1
847k                     1
562k                     1
54k                      1
                      ... 
13M                    191
14M                    194
12M                    196
11M                    198
Varies with device    1695
Name: count, Length: 462, dtype: int64

In [70]:
def convertSize(size):
  if 'M' in size:
    return np.float64(size.replace("M", ""))
  elif 'k' in size:
    return np.float64(size.replace('k', ""))/1024
  elif 'G' in size:
    return np.float64(size.replace('G', ""))*1024
  else:
    return np.nan
playstore['Size'] = playstore['Size'].apply(convertSize)

In [71]:
playstore.groupby("Category")["Size"].mean().sort_values(ascending=False)

Category
GAME                   44.370065
FAMILY                 28.319023
TRAVEL_AND_LOCAL       24.654463
SPORTS                 24.338822
HEALTH_AND_FITNESS     22.768851
PARENTING              22.512963
FOOD_AND_DRINK         22.185000
ENTERTAINMENT          21.853333
EDUCATION              20.790600
AUTO_AND_VEHICLES      20.036807
MEDICAL                19.628456
FINANCE                17.928309
SOCIAL                 17.681671
PHOTOGRAPHY            17.367574
DATING                 16.625555
MAPS_AND_NAVIGATION    16.614368
SHOPPING               16.175059
VIDEO_PLAYERS          16.084441
HOUSE_AND_HOME         15.978539
LIFESTYLE              14.754111
BUSINESS               14.471877
EVENTS                 13.963617
BEAUTY                 13.795745
COMICS                 13.484869
BOOKS_AND_REFERENCE    13.310188
WEATHER                13.124006
NEWS_AND_MAGAZINES     13.023615
PRODUCTIVITY           12.980885
ART_AND_DESIGN         12.370968
COMMUNICATION          11.765673
P

In [72]:
categorySizeMean = playstore.groupby('Category')['Size'].mean()

playstore['Size'] = playstore['Size'].fillna(playstore['Category'].map(categorySizeMean))

### Working with `Reviews` feature in **playstore** dataset

In [73]:
def convertReviews(review):
  if 'M' in review:
    return int(float(review.replace('M', ''))) * 1000000
  else:
    return np.int32(review)

playstore['Reviews'] = playstore['Reviews'].apply(convertReviews)

### Working with `Last Updated` feature in **playstore** dataset and aading new column `Year`

In [74]:
playstore['Last Updated'] = pd.to_datetime(playstore['Last Updated'], errors="coerce")

In [75]:
playstore["Year"] = playstore['Last Updated'].dt.year

### Fixing missing `Ratings` in **playstore** dataset

In [76]:
categoryRatingMedian = playstore.groupby('Category')['Rating'].median()

playstore['Rating'] = playstore['Rating'].fillna(playstore['Category'].map(categoryRatingMedian))

In [77]:
playstore["Rating"].value_counts()

Rating
4.3     1979
4.4     1250
4.2     1227
4.5     1154
4.6      823
4.1      747
4.0      568
4.7      499
3.9      386
3.8      303
5.0      274
3.7      239
4.8      234
3.6      174
3.5      163
3.4      128
3.3      102
4.9       87
3.0       83
3.1       69
3.2       64
2.9       45
2.8       42
2.6       25
2.7       25
2.5       21
2.3       20
2.4       19
1.0       16
2.2       14
1.9       13
2.0       12
1.7        8
1.8        8
2.1        8
1.6        4
1.4        3
1.5        3
1.2        1
19.0       1
Name: count, dtype: int64

### Working with *Outliers* and *NaN*

#### *Fixing PlayStore*

In [78]:
playstore = playstore[playstore['Category'] != '1.9']

In [79]:
playstore = playstore[playstore['Rating'] <= 5.0]

In [80]:
playstore.dropna(subset=['Rating', 'Type', 'Content Rating'], inplace=True)

In [81]:
playstore[['Android Ver', 'Current Ver']] = playstore[['Android Ver', 'Current Ver']].fillna('Unknown')

In [82]:
playstore = playstore.drop_duplicates()

In [83]:
playstore.isnull().sum()

App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Type              0
Price             0
Content Rating    0
Genres            0
Last Updated      0
Current Ver       0
Android Ver       0
Year              0
dtype: int64

In [84]:
playstore.duplicated().sum()

np.int64(0)

#### *Fixing UserReviews*

In [85]:
userreviews = userreviews.drop_duplicates()

In [86]:
playstore.isnull().sum()

App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Type              0
Price             0
Content Rating    0
Genres            0
Last Updated      0
Current Ver       0
Android Ver       0
Year              0
dtype: int64

In [87]:
playstore.duplicated().sum()

np.int64(0)

### **Conclusion**

At this phase we focused completely on addressing incoreect datatypes, handling missing values and duplicates. We often work with outliers and removed them. At this phase our primary focus was on `playstore` dataset.

- **Data Type Consistency:**  Critical columns like `Installs`, `Price`, `Size`, `Reviews`, and `Last Updated` have been converted to appropriate numerical types
- **Addressing missng values:** Handled missing values in `Size` and `Rating` via the *category-wise* mean and median.
- **Outliers Removed:** Removed data points, such as an invalid category and out-of-range ratings to improve quality data.
- **Duplicay Removal:** All duplicate entries in both datasets were removed, ensuring that each record contributes uniquely to the analysis.

In [88]:
playstore.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10356 entries, 0 to 10840
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   App             10356 non-null  object        
 1   Category        10356 non-null  object        
 2   Rating          10356 non-null  float64       
 3   Reviews         10356 non-null  int64         
 4   Size            10356 non-null  float64       
 5   Installs        10356 non-null  int64         
 6   Type            10356 non-null  object        
 7   Price           10356 non-null  float16       
 8   Content Rating  10356 non-null  object        
 9   Genres          10356 non-null  object        
 10  Last Updated    10356 non-null  datetime64[ns]
 11  Current Ver     10356 non-null  object        
 12  Android Ver     10356 non-null  object        
 13  Year            10356 non-null  float64       
dtypes: datetime64[ns](1), float16(1), float64(3), int64(2), obj

In [89]:
userreviews.info()

<class 'pandas.core.frame.DataFrame'>
Index: 30679 entries, 0 to 64236
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     30679 non-null  object 
 1   Translated_Review       29692 non-null  object 
 2   Sentiment               29697 non-null  object 
 3   Sentiment_Polarity      29697 non-null  float64
 4   Sentiment_Subjectivity  29697 non-null  float64
dtypes: float64(2), object(3)
memory usage: 1.4+ MB


In [90]:
playstore.to_csv("./clean/Play Store Data.csv", index=False)
userreviews.to_csv("./clean/User Reviews.csv", index=False)